<!--
TODO:
1. Allow output verbosity to be controlled from P3 settings
2. Create a wrapper R script which can be run from snakemake, sets relevant
   variables and knits the report.
3. Add support for coloring samples by a specified field.
-->

```{r, echo=FALSE, message=FALSE}
library('knitr')
library('ggplot2')
library('gplots')
library('reshape2')
library('RColorBrewer')
source('shared/visualization.R')

set.seed(1)

opts_chunk$set(fig.width=1080/96,
               fig.height=1080/96,
               dpi=96)
```

```{r output_header, results='asis', echo=FALSE}
header <- sprintf("P3: RNA-Seq (%s)", analysis_title)
cat(paste0(header, '\n', paste(rep('=', nchar(header)), collapse=''), '\n\n'))

cat(format(Sys.time(), "Generated: <time>%Y-%m-%d</time>\n"))
```

Overview
--------

```{r rnaseq, results='asis', echo=FALSE}
dat <- read.delim(infile, row.names=1)
mat <- as.matrix(dat)

cat(sprintf('\n- Number of samples: **%d**\n', ncol(dat)))
cat(sprintf('\n- Number of genes: **%d**\n', nrow(dat)))
cat(sprintf('\n- Mean # reads: **%0.2f**\n', mean(mat)))
cat(sprintf('\n- Median # reads: **%d**\n', median(mat)))
cat(sprintf('\n- Maximum # reads: **%d**\n', max(mat)))
```

Samples
-------

### Sample size

```{r rnaseq_sample_sizes, echo=FALSE}
sample_sizes <- data.frame(
    name=colnames(dat),
    size=colSums(dat)
)

ggplot(sample_sizes, aes(x=name, y=size)) + 
    geom_bar(stat="identity") +
    xlab("Sample") + ylab("Library size (# reads)") +
    theme(axis.text.x=element_text(angle=90, hjust=1))
```

### Count distributions

#### Log2 counts

```{r rnaseq_count_dists, echo=FALSE}
log2_counts <- melt(log2(mat + 1))
colnames(log2_counts) <- c('gene_id', 'sample', 'value')
ggplot(log2_counts, aes(x=value, color=sample)) + 
    geom_density() + 
    theme(legend.position="none")
```

#### Log2-CPM counts

```{r log2cpm_counts, echo=FALSE}
# Counts-per-million (CPM)
cpm <- function (x) {
    sweep(x, 2, colSums(x), '/') * 1E6
}
log2cpm_counts <- log2(cpm(mat) + 1)
log2cpm_counts_long <- melt(log2cpm_counts)

colnames(log2cpm_counts_long) <- c('gene_id', 'sample', 'value')

ggplot(log2cpm_counts_long, aes(x=value, color=sample)) + 
    geom_density() +
    theme(legend.position="none")
```

### Samples

#### Sample heatmap (euclidean distance)

```{r sample_heatmap_dist, echo=FALSE}
heatmap_colors <- rev(colorRampPalette(brewer.pal(9, "YlGnBu"))(100))
heatmap.2(as.matrix(dist(t(mat))), trace='none', col=heatmap_colors)
```

#### Sample heatmap (pearson correlation)

```{r sample_heatmap_cor_pearson, echo=FALSE}
heatmap.2(cor(mat), trace='none')
```

#### Sample Heatmap (spearman correlation)

```{r sample_heatmap_cor_spearman, echo=FALSE}
heatmap.2(cor(mat, method='spearman'), trace='none')
```

#### Sample PCA plot

```{r sample_pca, echo=FALSE}
# Filter zero-variance genes
mat_var_filtered <- mat[apply(mat, 1, var) != 0,]
plot_sample_pca(mat_var_filtered)
```

Genes
-----

### Log2-CPM counts (low-count filtered)

```{r rnaseq_gene_heatmap, echo=FALSE}
# filter out low-count genes 
log2cpm_counts_filtered <- filter_low_counts(log2cpm_counts)

# sub-sample data
ind <- sample(1:nrow(log2cpm_counts_filtered), 2000)
heatmap.2(log2cpm_counts_filtered[ind,], trace='none', labRow=FALSE,
          col=redgreen(75))
```

System info
-----------

```{r, echo=FALSE}
date()
sessionInfo()
```
